### PR TITLE
add end to end test for function listUsersCmdF

### PR DESCRIPTION
**Pull request by M-Buntoro**
Summary

Add end to end test case for command mmctl user list

Ticket Link

https://mattermost.atlassian.net/browse/MM-29153
Fixes https://github.com/mattermost/mattermost-server/issues/15663

### CODE REVIEW

File: commands/user_e2e_test.go [hunk 1]
```diff
@@ -6,6 +6,8 @@ package commands
 import (
 	"github.com/mattermost/mattermost-server/v5/model"
 	"github.com/spf13/cobra"
+	// "github.com/spf13/pflag"
+	"log"
```
Comment (hanzei) [inline commands/user_e2e_test.go:10]: You need to order the import differently to fix CI:
```suggestion
	"log"
	
	"github.com/mattermost/mattermost-server/v5/model"
	"github.com/spf13/cobra"
```

File: commands/user_e2e_test.go [hunk 2]
```diff
@@ -136,3 +138,63 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		usr := s.th.CreateUser()
+		userPool[usr.Username] = true
+	}
+
+	s.RunForAllClients("Get some random of user", func(c client.Client) {
+		printer.Clean()
+
+		page:= int(0)
+		perpage:= int(5)
+		all := bool(false)
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:162]: No need to cast or even assign for zero values.
```suggestion
		var page int
		perpage := 5
		var all bool
```

File: commands/user_e2e_test.go [hunk 3]
```diff
@@ -136,3 +138,63 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+
+		err := listUsersCmdF(c, cmd, []string{})
+		s.Require().Nil(err)
+		s.Len(printer.GetLines(), 5)
+		s.Len(printer.GetErrorLines(), 0)
+		log.Println(printer.GetErrorLines())
+
+		for _,each := range printer.GetLines() {
+			eachUser := each.(*model.User)
+			s.Require().Equal(userPool[eachUser.Username], true)
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:176]: I'd prefer more convenient variables names, and also we can use `require.Contains` here.
```suggestion
		for _, u := range printer.GetLines() {
			user := u.(*model.User)
			s.Require().Contains(userPool, user.Username)
```

File: commands/user_e2e_test.go [hunk 4]
```diff
@@ -136,3 +138,63 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		}
+	})
+
+	s.RunForAllClients("Get list of all user", func(c client.Client) {
+		printer.Clean()
+
+		page:= int(0)
+		perpage:= int(10)
+		all := bool(true)
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:185]: Same here.

File: commands/user_e2e_test.go [hunk 5]
```diff
@@ -136,3 +138,63 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		perpage:= int(10)
+		all := bool(true)		
+		cmd := &cobra.Command{}
+		cmd.Flags().IntVar(&page,"page", page,"page")
+		cmd.Flags().IntVar(&perpage,"per-page", perpage,"perpage")
+		cmd.Flags().BoolVar(&all, "all",all,"all")
+
+		err := listUsersCmdF(c, cmd, []string{})
+		s.Require().Nil(err)
+		s.Len(printer.GetLines(), 14)
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:193]: I wouldn't be so certain if our test suite will produce `4` users every time. Maybe we can use `GreaterOrEqual`?

File: commands/user_e2e_test.go [hunk 6]
```diff
@@ -136,3 +138,63 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		s.Len(printer.GetErrorLines(), 0)
+		for _,each := range printer.GetLines() {
+			eachUser := each.(*model.User)
+			s.Require().Equal(userPool[eachUser.Username], true)
+		}
+	})
+}
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:200]:
```suggestion
}

```

File: commands/user_e2e_test.go [hunk 7]
```diff
@@ -136,3 +138,63 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		cmd.Flags().IntVar(&perpage,"per-page", perpage,"perpage")
+		cmd.Flags().BoolVar(&all, "all",all,"all")
+
+		err := listUsersCmdF(c, cmd, []string{})
+		s.Require().Nil(err)
+		s.Len(printer.GetLines(), 14)
+		s.Len(printer.GetErrorLines(), 0)
+		for _,each := range printer.GetLines() {
+			eachUser := each.(*model.User)
+			s.Require().Equal(userPool[eachUser.Username], true)
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:197]: Same as above.

**Review by isacikgoz — CHANGES_REQUESTED** [review]
Thanks again @M-Buntoro very nice implementation 👍 I'd like to request some minor changes for overall consistency.

File: commands/user_e2e_test.go (hunk 5 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:192]: cool, I'll change it to greater or equal

File: commands/user_e2e_test.go (hunk 3 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:175]: sure

File: commands/user_e2e_test.go [hunk 8]
```diff
@@ -136,3 +136,62 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
 		s.Equal("Unable to find user '"+emailArg+"'", printer.GetErrorLines()[0])
 	})
 }
+
+func (s *MmctlE2ETestSuite) TestListUserCmd() {
```
Comment (hanzei) [inline commands/user_e2e_test.go:140]: The tests fail on CI:
https://app.circleci.com/pipelines/github/mattermost/mmctl/1566/workflows/40e58ca7-452e-4a71-8133-a78352baca2e/jobs/5397/parallel-runs/0/steps/0-105

Could you please take a look what is going on?

**Review by hanzei — DISMISSED** [review]

File: commands/user_e2e_test.go (hunk 8 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:140]: on it

Comment (M-Buntoro) [reply commands/user_e2e_test.go:140]: sorry, I forgot to add len on the asssertion, already adjusted
thanks @hanzei :+1:

File: commands/user_e2e_test.go [hunk 9]
```diff
@@ -136,3 +136,62 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+
+	// populate map for checking
+	userPool := []string{
+		s.th.BasicUser.Username,
+		s.th.BasicUser2.Username,
+		s.th.TeamAdminUser.Username,
+		s.th.SystemAdminUser.Username,
+	}
+	for i := 0; i < 10; i++ {
+		usr := s.th.CreateUser()
```
Comment (mgdelacroix) [inline commands/user_e2e_test.go:151]: nit: just to make sure that user creation goes well, could you please add an error check here?

```suggestion
		usr, err := s.th.CreateUser()
		s.Require().Nil(err)
```

File: commands/user_e2e_test.go [hunk 10]
```diff
@@ -136,3 +136,62 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		var all bool
+		perpage := 5
+		cmd := &cobra.Command{}
+		cmd.Flags().IntVar(&page, "page", page, "page")
+		cmd.Flags().IntVar(&perpage, "per-page", perpage, "perpage")
+		cmd.Flags().BoolVar(&all, "all", all, "all")
+
+		err := listUsersCmdF(c, cmd, []string{})
+		s.Require().Nil(err)
+		s.Require().GreaterOrEqual(len(printer.GetLines()), 5)
```
Comment (mgdelacroix) [inline commands/user_e2e_test.go:168]: Why the greater or equal here instead of just an `Len` check?

File: commands/user_e2e_test.go [hunk 11]
```diff
@@ -136,3 +136,62 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		perpage := 10
+		all := true
+		cmd := &cobra.Command{}
+		cmd.Flags().IntVar(&page, "page", page, "page")
+		cmd.Flags().IntVar(&perpage, "per-page", perpage, "perpage")
+		cmd.Flags().BoolVar(&all, "all", all, "all")
+
+		err := listUsersCmdF(c, cmd, []string{})
+		s.Require().Nil(err)
+		s.Require().GreaterOrEqual(len(printer.GetLines()), 14)
```
Comment (mgdelacroix) [inline commands/user_e2e_test.go:190]: Same `GreaterOrEqual` question here

File: commands/user_e2e_test.go (hunk 10 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:168]: it was suggested by @isacikgoz as the number of user created on initialization could increase from time to time, and may not be static

File: commands/user_e2e_test.go (hunk 11 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:191]: Same as above

File: commands/user_e2e_test.go (hunk 10 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:169]: @mgdelacroix should I change to standard `Len` check for this case?

**Review by isacikgoz — APPROVED** [review]
Nicely done, thanks for the contribution @M-Buntoro 🎉

File: commands/user_e2e_test.go (hunk 10 above)
Comment (isacikgoz) [reply commands/user_e2e_test.go:174]: I am 0/5 on this. Either is fine actually.

Comment (mgdelacroix) [reply commands/user_e2e_test.go:174]: Apologies for the delay. Understood :)

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1: Thanks a lot @M-Buntoro!

**Review by lindy65 — APPROVED** [review]
Thanks @M-Buntoro @mgdelacroix - looks good to me 👍

**Pull request merged**